File: transcriber.py
import logging
import httpx
from soniox import SonioxClient
from soniox.types import CreateTranscriptionConfig
from soniox.utils import render_tokens

logger = logging.getLogger(__name__)


class Transcriber:
    def __init__(self, soniox_api_key: str):
        self._client = SonioxClient(api_key=soniox_api_key)

    def transcribe_voice(self, telegram_token: str, file_id: str) -> str | None:
        """Download a Telegram voice file and transcribe it via Soniox."""
        # Get file path from Telegram
        with httpx.Client() as http:
            resp = http.get(f"https://api.telegram.org/bot{telegram_token}/getFile",
                            params={"file_id": file_id})
            file_path = resp.json()["result"]["file_path"]

            # Download the voice file
            resp = http.get(f"https://api.telegram.org/file/bot{telegram_token}/{file_path}")
            audio_bytes = resp.content

        logger.info("Downloaded voice file: %s (%d bytes)", file_path, len(audio_bytes))

        # Upload to Soniox
        uploaded = self._client.files.upload(audio_bytes)
        logger.info("Uploaded to Soniox: file_id=%s", uploaded.id)

        try:
            # Create transcription
            transcription = self._client.transcriptions.create(
                config=CreateTranscriptionConfig(
                    model="stt-async-v4",
                    language_hints=["es", "en"],
                ),
                file_id=uploaded.id,
            )

            # Wait for completion
            self._client.transcriptions.wait(transcription.id)

            # Get result
            result = self._client.transcriptions.get_transcript(transcription.id)
            text = render_tokens(result.tokens, []).strip()
            logger.info("Transcription complete: %r", text[:120])

            # Cleanup
            self._client.transcriptions.delete(transcription.id)

            return text if text else None
        finally:
            self._client.files.delete(uploaded.id)

File: src/services/transcriber.ts
import { SonioxNodeClient } from "@soniox/node";
import { createLogger } from "../logger.js";

const log = createLogger("transcriber");

/**
 * Transcribe a Telegram voice note using the Soniox async STT API.
 *
 * Flow:
 *  1. Resolve the file path via Telegram's getFile endpoint.
 *  2. Download the audio bytes from Telegram's file server.
 *  3. Upload to Soniox, transcribe with Spanish/English hints, and wait.
 *  4. Return the transcript text (or null when empty / on failure).
 */
export async function transcribeVoice(
  botToken: string,
  sonioxApiKey: string,
  fileId: string,
): Promise<string | null> {
  // 1. Get the file path from Telegram
  const fileRes = await fetch(`https://api.telegram.org/bot${botToken}/getFile?file_id=${fileId}`);
  if (!fileRes.ok) {
    log.error({ status: fileRes.status }, "Failed to get file from Telegram");
    return null;
  }

  const fileData = (await fileRes.json()) as {
    ok: boolean;
    result?: { file_path?: string };
  };
  const filePath = fileData.result?.file_path;
  if (!filePath) {
    log.error({ fileData }, "Telegram getFile returned no file_path");
    return null;
  }

  // 2. Download the audio from Telegram
  const audioRes = await fetch(`https://api.telegram.org/file/bot${botToken}/${filePath}`);
  if (!audioRes.ok) {
    log.error({ status: audioRes.status }, "Failed to download audio file");
    return null;
  }

  const audioBuffer = Buffer.from(await audioRes.arrayBuffer());
  log.debug({ bytes: audioBuffer.length, filePath }, "Downloaded voice file");

  // 3. Upload to Soniox and transcribe
  const client = new SonioxNodeClient({ api_key: sonioxApiKey });

  const transcription = await client.stt.transcribe({
    model: "stt-async-preview",
    file: audioBuffer,
    filename: filePath.split("/").pop() ?? "voice.ogg",
    language_hints: ["es", "en"],
    wait: true,
    cleanup: ["file", "transcription"],
  });

  const transcript = transcription.transcript;
  const text = transcript?.text?.trim() ?? null;

  if (!text) {
    log.warn("Soniox returned empty transcript");
    return null;
  }

  log.debug({ chars: text.length }, "Transcription complete");
  return text;
}
